File: SKUDataFilterDemo/SKUDataFilterDemo/ViewController.m
//
//  ViewController.m
//  SKUDataFilterDemo
//
//  Created by OrangesAL on 2017/12/4.
//  Copyright © 2017年 OrangesAL. All rights reserved.
//

#import "ViewController.h"
#import "ORSKUDataFilter.h"

@interface ViewController ()<UICollectionViewDataSource, UICollectionViewDelegate,ORSKUDataFilterDataSource>

@property (weak, nonatomic) IBOutlet UILabel *priceL;
@property (weak, nonatomic) IBOutlet UILabel *storeL;
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;

@property (nonatomic, strong) NSArray *dataSource;
@property (nonatomic, strong) NSArray *skuData;;
@property (nonatomic, strong) NSMutableArray <NSIndexPath *>*selectedIndexPaths;;

@property (nonatomic, strong) ORSKUDataFilter *filter;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    
    NSArray *dataSource = @[@{@"name" : @"款式",
                              @"value" : @[@"male", @"famale"]},
                            @{@"name" : @"颜色",
                              @"value" : @[@"red", @"green", @"blue"]},
                            @{@"name" : @"尺寸",
                              @"value" : @[@"XXL", @"XL", @"L", @"S", @"M"]},
                            @{@"name" : @"test",
                              @"value" : @[@"A", @"B", @"C"]},];
    _dataSource = dataSource;
    
    _selectedIndexPaths = [NSMutableArray array];
    
    _skuData = @[
                 @{@"contition":@"male,red,XL,A",
                   @"price":@"1120",
                   @"store":@"167"},
                 @{@"contition":@"male,red,M,B",
                   @"price":@"1200",
                   @"store":@"289"},
                 @{@"contition":@"male,green,L,A",
                   @"price":@"889",
                   @"store":@"300"},
                 @{@"contition":@"male,green,M,B",
                   @"price":@"991",
                   @"store":@"178"},
                 @{@"contition":@"famale,red,XL,A",
                   @"price":@"1000",
                   @"store":@"200"},
                 @{@"contition":@"famale,blue,L,B",
                   @"price":@"880",
                   @"store":@"12"},
                 @{@"contition":@"famale,blue,XXL,C",
                   @"price":@"1210",
                   @"store":@"300"},
                 @{@"contition":@"male,blue,L,C",
                   @"price":@"888",
                   @"store":@"121"},
                 @{@"contition":@"famale,green,M,C",
                   @"price":@"1288",
                   @"store":@"125"},
                 @{@"contition":@"male,blue,L,A",
                   @"price":@"1210",
                   @"store":@"123"}
                 ];
    
    _filter = [[ORSKUDataFilter alloc] initWithDataSource:self];
    
    //当数据更新的时候 需要reloadData
    [_filter reloadData];
    [self.collectionView reloadData];
    
}

#pragma mark -- collectionView

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return _dataSource.count;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [_dataSource[section][@"value"] count];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
        
    PropertyCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"PropertyCell" forIndexPath:indexPath];
    NSArray *data = _dataSource[indexPath.section][@"value"];
    cell.propertyL.text = data[indexPath.row];
    
    if ([_filter.availableIndexPathsSet containsObject:indexPath]) {
        [cell setTintStyleColor:[UIColor blackColor]];
    }else {
        [cell setTintStyleColor:[UIColor lightGrayColor]];
    }
    
    if ([_filter.selectedIndexPaths containsObject:indexPath]) {
        [cell setTintStyleColor:[UIColor redColor]];
    }
    return cell;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
    
    if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
        PropertyHeader *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"headerIdf" forIndexPath:indexPath];
        view.headernameL.text = _dataSource[indexPath.section][@"name"];
        return view;

    }else {
        UICollectionReusableView *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"footerIdf" forIndexPath:indexPath];
        
        return view;
    }
    
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    [_filter didSelectedPropertyWithIndexPath:indexPath];
    
    [collectionView reloadData];
    [self action_complete:nil];
}

#pragma mark -- ORSKUDataFilterDataSource

- (NSInteger)numberOfSectionsForPropertiesInFilter:(ORSKUDataFilter *)filter {
    return _dataSource.count;
}

- (NSArray *)filter:(ORSKUDataFilter *)filter propertiesInSection:(NSInteger)section {
    return _dataSource[section][@"value"];
}

- (NSInteger)numberOfConditionsInFilter:(ORSKUDataFilter *)filter {
    return _skuData.count;
}

- (NSArray *)filter:(ORSKUDataFilter *)filter conditionForRow:(NSInteger)row {
    NSString *condition = _skuData[row][@"contition"];
    return [condition componentsSeparatedByString:@","];
}

- (id)filter:(ORSKUDataFilter *)filter resultOfConditionForRow:(NSInteger)row {
    NSDictionary *dic = _skuData[row];
    return @{@"price": dic[@"price"],
             @"store": dic[@"store"]};
}

#pragma mark -- action
- (IBAction)action_complete:(id)sender {
    
    
    NSDictionary *dic = _filter.currentResult;
    
    if (dic == nil) {
        NSLog(@"请选择完整 属性");
        _priceL.text = @"￥0";
        _storeL.text = @"库存0件";
        return;
    }
    
    _priceL.text = [NSString stringWithFormat:@"￥%@",dic[@"price"]];
    _storeL.text = [NSString stringWithFormat:@"库存%@件",dic[@"store"]];
}



@end


@implementation PropertyCell

- (void)setTintStyleColor:(UIColor *)color {
    self.layer.borderColor = color.CGColor;
    self.propertyL.textColor = color;
}

@end

@implementation PropertyHeader

@end

File: SKUDataFilterDemo-Swift/SKUDataFilterDemo-Swift/ViewController.swift
//
//  ViewController.swift
//  SKUDataFilterDemo-Swift
//
//  Created by 欧阳荣 on 2019/9/8.
//  Copyright © 2019 OrangesAL. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var priceL: UILabel!
    @IBOutlet weak var storeL: UILabel!
    @IBOutlet weak var collectionView: UICollectionView!

    private let dataSource = [
    
        ["name" : "款式",
         "value" : ["male", "famale"]],
        
        ["name" : "颜色",
         "value" : ["red", "green", "blue"]],
        
        ["name" : "尺寸",
         "value" : ["XXL", "XL", "L", "S", "M"]],
        
        ["name" : "test",
         "value" : ["A", "B", "C"]]
    ]
    
    private let skuData : [[String : String]] = [
        
        ["condition" : "male,red,XL,A",
         "price" : "1120",
         "store" : "167"],
        
        ["condition" : "male,red,M,B",
         "price" : "1200",
         "store" : "289"],
        
        ["condition" : "male,green,L,A",
         "price" : "889",
         "store" : "300"],
        
        ["condition" : "male,green,M,B",
         "price" : "991",
         "store" : "178"],
        
        ["condition" : "famale,red,XL,A",
         "price" : "1000",
         "store" : "200"],
        
        ["condition" : "famale,blue,L,B",
         "price" : "880",
         "store" : "12"],
        
        ["condition" : "famale,blue,XXL,C",
         "price" : "1210",
         "store" : "300"],
        
        ["condition" : "male,blue,L,C",
         "price" : "888",
         "store" : "121"],
        
        ["condition" : "famale,green,M,C",
         "price" : "1288",
         "store" : "125"],
        
        ["condition" : "male,blue,L,A",
         "price" : "1210",
         "store" : "123"]
    ]
    
    private lazy var filter = ORSKUDataFilter(dataSource: self);
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        filter.needDefaultValue = true
        collectionView.reloadData()
        self.updateResult()
    }
    
    func updateResult() {
        
        guard let dic = filter.currentResult as? [String : String] else {
            self.priceL.text = "￥0";
            self.storeL.text = "库存0件";
            print("属性不完整")
            return
        }
        
        priceL.text = "￥" + (dic["price"] ?? "0")
        storeL.text = "库存" + (dic["store"] ?? "0") + "件"
    }
    
    @IBAction func action_complete(_ sender: Any) {
        self.updateResult()
    }
    
}


extension ViewController: UICollectionViewDataSource, UICollectionViewDelegate {
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return dataSource.count
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        
        let value = dataSource[section]["value"] as! Array<Any>
        return value.count;
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(
            withReuseIdentifier: "PropertyCell",
            for: indexPath
            ) as! PropertyCell
        
        let value = dataSource[indexPath.section]["value"] as! Array<String>
        cell.propertyL.text = value[indexPath.row]
        
        if filter.selectedIndexPaths.contains(indexPath) {
            cell.tintStyleColor = UIColor.red
        }else if filter.availableIndexPathsSet.contains(indexPath) {
            cell.tintStyleColor = UIColor.black
        }else {
            cell.tintStyleColor = UIColor.lightGray
        }
        
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        
        if kind == UICollectionView.elementKindSectionHeader {
            let view = collectionView.dequeueReusableSupplementaryView(
                ofKind: kind,
                withReuseIdentifier: "headerIdf",
                for: indexPath
                ) as! PropertyHeader
            
            let title = dataSource[indexPath.section]["name"] as! String
            view.headernameL.text = title;
            return view
        }
        
        let view = collectionView.dequeueReusableSupplementaryView(
            ofKind: kind,
            withReuseIdentifier: "footerIdf",
            for: indexPath
            )
        return view
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        filter.selectedProperty(at: indexPath)
        collectionView.reloadData()
        self.updateResult()
    }
    
}

extension ViewController : ORSKUDataFilterDataSource {
    
    func numberOfSectionsForProperties(filter: ORSKUDataFilter) -> Int {
        return dataSource.count
    }
    
    func properties(filter: ORSKUDataFilter, in section: Int) -> [AnyHashable] {
        return dataSource[section]["value"] as! [AnyHashable]
    }
    
    func numberOfConditions(filter: ORSKUDataFilter) -> Int {
        return skuData.count
    }
    
    func condition(filter: ORSKUDataFilter, at row: Int) -> [AnyHashable] {
        
        let skuString = skuData[row]["condition"]
        return skuString!.components(separatedBy: ",")
    }
    
    func resultOfCondition(filter: ORSKUDataFilter, at row: Int) -> Any? {
        return skuData[row]
    }
    
    
}

class PropertyCell: UICollectionViewCell {
    
    @IBOutlet weak var propertyL: UILabel!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        self.layer.borderWidth = 1
    }
    
    var tintStyleColor : UIColor? {
        set {
            self.layer.borderColor = newValue?.cgColor
            self.propertyL.textColor = newValue
        }
        
        get {
            return self.propertyL.textColor
        }
    }
    
}

class PropertyHeader: UICollectionReusableView {
    
    @IBOutlet weak var headernameL: UILabel!
}
